### PR TITLE
Selective record

### PR DESCRIPTION
**Pull request by Christian-B**
Includes https://github.com/SpiNNakerManchester/sPyNNaker/pull/445 and its partners.

Selectiing recording by sampling rate but not on spikes yet.

### CODE REVIEW

File: spynnaker/pyNN/models/recording_common.py [hunk 1]
```diff
@@ -101,12 +103,39 @@ def _set_spikes_recording(self):
...
+        if variable == "spikes":
+            data = self._get_spikes()
+
+        (data, ids, sampling_interval) = self._get_recorded_matrix(variable)
+        return self.pynn7_format(data, ids, sampling_interval)
+
+    def _get_recorded_matrix(self, variable):
         """ method that contains all the safety checks and gets the recorded
-        data from the vertex
+        data from the vertex in matrix format
```
Comment (rowleya) [inline spynnaker/pyNN/models/recording_common.py]: indent me!

File: spynnaker/pyNN/models/recording_common.py [hunk 2]
```diff
@@ -101,12 +103,39 @@ def _set_spikes_recording(self):
...
 
         :param variable: the variable name to read. supported variable names
-        are :'gsyn_exc', 'gsyn_inh', 'v', 'spikes'
+        are :'gsyn_exc', 'gsyn_inh', 'v'
```
Comment (rowleya) [inline spynnaker/pyNN/models/recording_common.py]: indent me!

File: neural_modelling/src/neuron/neuron.c [hunk 3]
```diff
@@ -52,9 +52,30 @@ static uint32_t n_neurons;
 //! The recording flags
 static uint32_t recording_flags;
 
+//static uint32_t cyab_counter;
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply neural_modelling/src/neuron/neuron.c]: good catch done in later branch

File: neural_modelling/src/neuron/neuron.c [hunk 4]
```diff
@@ -90,7 +111,7 @@ typedef enum parmeters_in_neuron_parameter_data_region {
 static inline void _print_neurons() {
 
 //! only if the models are compiled in debug mode will this method contain
-//! said lines.
+//! said lines.neuron_t
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply neural_modelling/src/neuron/neuron.c]: good catch removed in later branch

File: spynnaker/pyNN/models/common/neuron_recorder.py [hunk 5]
```diff
@@ -1,58 +1,209 @@
...
+        :param variable: PyNN name of the variable
+        :return: Sampling interval in micro seconds
+        """
+        return recording_utils.compute_interval(self._sampling_rates[variable])
+
+    def get_matrix_data(
+            self, label, buffer_manager, region, placements, graph_mapper,
+            application_vertex, variable, n_machine_time_steps):
+        """ method for reading a uint32 mapped to time and neuron ids from\
+        the SpiNNaker machine
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/common/neuron_recorder.py]: good catch done in later branch

File: spynnaker/pyNN/models/common/neuron_recorder.py [hunk 6]
```diff
@@ -1,58 +1,209 @@
...
+            record_raw = neuron_param_region_data_pointer.read_all()
+            record_length = len(record_raw)
+            row_length = (n_neurons + 1) * self.N_BYTES_PER_NEURON
+            # There is one column for time and one for each neuron recording
+            n_rows = record_length // row_length
+            # Converts bytes to ints and make a matrix
+            record = (numpy.asarray(record_raw, dtype="uint8").
+                      view(dtype="<i4")).reshape((n_rows, (n_neurons + 1)))
+            # Check if you have the expected data
+            if missing_data or n_rows != expected_rows:
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/common/neuron_recorder.py]: Agreed In theory no that is not possible.

But it is good safety code at a minor cost that has helped me more than once, while tweaking the sampling stuff.

File: spynnaker/pyNN/models/common/neuron_recorder.py [hunk 7]
```diff
@@ -1,58 +1,209 @@
...
+                    time = i * sampling_interval
+                    # Check if there is data for this timestep
+                    indexes = numpy.where(record[:, 0] == time)
+                    if len(indexes[0]) > 0:
+                        # Set row to data for that timestep
+                        fragment[i] = record[indexes[0][0], 1:]
+                    else:
+                        # Set row to nan
+                        fragment[i] = numpy.full(n_neurons, numpy.nan)
+            else:
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/common/neuron_recorder.py]: Done in later branch

File: spynnaker/pyNN/models/common/neuron_recorder.py [hunk 8]
```diff
@@ -1,58 +1,209 @@
...
             msg = "Variable {} is not supported ".format(variable)
             raise fec_excceptions.ConfigurationException(msg)
 
-    def get_sdram_usage_in_bytes(self, variable, n_neurons,
-                                 n_machine_time_steps):
-        if self.is_recording(variable):
-            return recording_utils.get_recording_region_size_in_bytes(
-                n_machine_time_steps,  self.N_BYTES_PER_NEURON * n_neurons)
-        else:
+    def _count_recording_per_slice(self, variable, slice):
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/common/neuron_recorder.py]: Agreed fixed in later branch

File: spynnaker/pyNN/models/common/neuron_recorder.py [hunk 9]
```diff
@@ -1,58 +1,209 @@
...
+        if self._sampling_rates[variable] == 0:
             return 0
+        return slice.n_atoms
+
+    def get_buffered_sdram_per_timestep(self, variable, slice):
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/common/neuron_recorder.py]: Agreed fixed in later branch

File: spynnaker/pyNN/models/common/neuron_recorder.py [hunk 10]
```diff
@@ -1,58 +1,209 @@
...
+        :return:
+        """
+        neuron_count = self._count_recording_per_slice(variable, slice)
+        if neuron_count == 0:
+            return 0
+        rate = self._sampling_rates[variable]
+        data_size = self.N_BYTES_PER_NEURON * neuron_count
+        return (data_size + self.N_BYTES_FOR_TIMESTAMP) / rate
+
+    def get_extra_buffered_sdram(self, variable, slice):
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/common/neuron_recorder.py]: Agreed fixed in later branch

File: spynnaker/pyNN/models/common/neuron_recorder.py [hunk 11]
```diff
@@ -1,58 +1,209 @@
...
+
+        :param variable:
+        :param slice:
+        :return:
+        """
+        rate = self._sampling_rates[variable]
+        if rate <= 1:
+            # No sampling so get_buffered_sdram_per_timestep was correct
+            return 0
+        neuron_count = self._count_recording_per_slice(variable, slice)
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/common/neuron_recorder.py]: 1. This code has been removed in a later branch which uses a different approach.

2. rate == 0 signals NOT recording so no SDRAM is ever needed

File: spynnaker/pyNN/models/common/neuron_recorder.py [hunk 12]
```diff
@@ -1,58 +1,209 @@
...
 
     def get_n_cpu_cycles(self, n_neurons):
         return n_neurons * self.N_CPU_CYCLES_PER_NEURON * \
-                sum(self._record.values())
+                len(self.recording_variables)
+
+    def get_sdram_usage_for_global_parameters_in_bytes(self):
+        return len(self._sampling_rates) * 4
+
+    def get_global_parameters(self, slice):
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/common/neuron_recorder.py]: Agreed fixed in later branch

File: spynnaker/pyNN/models/common/neuron_recorder.py [hunk 13]
```diff
@@ -1,58 +1,209 @@
...
+        params = []
+        for variable in self._sampling_rates:
+            params.append(recording_utils.rate_parameter(
+                self._sampling_rates[variable]))
+        return params
+
+    def get_indexes_for_slice(self, variable, slice):
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/common/neuron_recorder.py]: Agreed fixed in later branch

File: spynnaker/pyNN/models/common/recording_utils.py [hunk 14]
```diff
@@ -90,3 +98,47 @@ def get_buffer_sizes(buffer_max, space_needed, enable_buffered_recording):
     if buffer_max < space_needed:
         return buffer_max
     return space_needed
+
+
+def compute_rate(new_state, sampling_interval):
+    """
+    Converts a simpling interval into a rate
```
[the remark this answers is not in the document]
Comment (Christian-B) [reply spynnaker/pyNN/models/common/recording_utils.py]: Agreed fixed in later branch

File: spynnaker/pyNN/models/recording_common.py (hunk 1 above)
Comment (Christian-B) [reply spynnaker/pyNN/models/recording_common.py]: Agreed fixed in later branch

File: spynnaker/pyNN/models/recording_common.py (hunk 2 above)
Comment (Christian-B) [reply spynnaker/pyNN/models/recording_common.py]: Agreed fixed in later branch

**Comment by Christian-B** [comment]
part of https://github.com/SpiNNakerManchester/sPyNNaker/pull/461

**Pull request closed without merging**